### DISCUSSION TITLE
Pipe operator

### DISCUSSION THREAD

**Discussion by astrale-sharp** [typst/typst#530] [Ideas]
# Why
Currently Typst relies heavily on functions. Nesting functions come with a cumbersome syntax at times:
```
function(arg1,other_func(arg2,arg3,last_func(arg4,arg5),arg6,arg7))
```

The pipe operator existing in other languages would permit the write the following in a more expressive way:

```
last_func(arg4,arg5) |> other_func(arg2,arg3, _ ,arg6,arg7) |> function(arg1, _)
```

# The specifics
Each of the following syntax sugar are considered :
- `arg |> func ` --> `func(arg)`
- `arg |> func(_) ` --> `func(arg)`
- `(arg1, arg2) |> func ` --> `func(arg1, arg2)`
- `(arg1, arg2) |> func( _, value, _) ` --> `func(arg1, value, arg2)`
- `..list |> func ` --> `func(..list)`
- `..dictionary |> func ` --> `func(key1: value1, key2: value2, ...)`

Not all of these should necessarily be implemented, I'm curious what's the community opinion on this, personally I'd like them all.

Opinions, Ideas, Alternatives?

**Comment by CGMossa** [comment]
Super controversial opinion, that I don't necessarily agree with. But we could have an operator that only work on things that passes a body or content, and picks up just that.
Meaning it will make it easy to insert stacks of modifications on top of each other and take them out with commenting, but it doesn't add any sugar for regular function composition.

And this could be like `.with`.

**Reply by astrale-sharp** [reply]
Thanks for your comment!
I'm not sure I understand all the ramifications, could you provide small examples?

**Comment by dianegolfouse** [comment]
> - (arg1, arg2) |> func --> func(arg1, arg2)
> - (arg1, arg2) |> func( _, value, _) --> func(arg1, value, arg2)

I think this would conflict with array syntax, no? I don't think it can differentiate between passing `(arg1, arg2)` as a single argument array or as two arguments.

Maybe the `..` syntax could be reused here : `..(arg1, arg2) |> func` --> `func(arg1, arg2)` ? Or something else...

**Reply by astrale-sharp** [reply]
Good point, that's a great idea

**Comment by idoric** [comment]
Couldn't we reuse the sign => of unamed functions, adding new possibilities to it and allowing it to be used in chains?

Edit: in the reference documentation it is said "You can also created an unnamed function without creating a binding by specifying a parameter list followed by => and the function body. If your function has just one parameter, the parentheses around the parameter list are optional."
So it is already possible to pass several parameters to an unamed function (although I can't find any concrete example in this documentation), which brings us closer to what is sought to be done.

**Reply by astrale-sharp** [reply]
Thanks for your comment!
 
 `a => a` is this equivalent to `a(a)` if it's piped or `f(x)=x` if it's a closure?

**Comment by enseignantePC** [comment]


**Reply by astrale-sharp** [reply]
I just thought again of an alternative we discussed a while back, instead of having all type have a method `pipe`, it could be just the functions. Or simply an `and` method
 
 ```rust
 // I want a box aligned right with inside a text with a specific font size aligned center + 
 
 #{
 // Possible now, my least favorite way, hard to read and write.
   align(right,box(align( center, text(size : 14pt,[Lorem Ipsum]))))
 
 // Possible now, slightly better but a lot of [#...] which makes it still hard to read.
   align(right)[#box[#align( center)[#text(size : 14pt)[Lorem Ipsum]]]]
 // Possible with `and` method, I think it's quite good. but I have to read it backwards, the start is on the right and then each operation take you to the left, which is why I prefer the pipe. 
 text.with(size: 14pt).and(align.with(center)).and(box).and(align.with(right))[Lorem Ipsum]
 // also possible with `and` method, I really don't like that one,
 text.with(size: 14pt)[Lorem Ipsum].and(align.with(center)).and(box).and(align.with(right))()
 // possible with the pipe method
 // I like the reading from left to right but I don't like ending with `()` which is required because of the `.with` method call
 text.with(size : 14pt)[Lorem Ipsum].pipe(align.with(center)).pipe(box).pipe(align.with(right))()
 // workaround so that you don't need `()` in the end
 let f() = [Lorem Ipsum]
 f.pipe(text.with(size:14pt)).pipe(align.with(center)).pipe(box).pipe(align.with(right))
 // possible with the pipe operator
 // My favourite because the number of parenthesis goes down drastically.
 [Lorem Ipsum] |> text.with(size:14pt) |> align.with(center) |> box |> align.with(right)
 // Also note that it's really .with heavy which I'm not sure how I feel about right now 
 }
 ```

**Comment by astrale-sharp** [comment]
One the discord there has been suggestions : 
```arg2 |> F(arg1)``` becomes `F(arg1,arg2)`
What are the preferences here?

**Reply by astrale-sharp** [reply]
Another path I see possible is 
`..(b,d) |> f(a,_,c,_)`  becomes `f(a,b,c,d)`
`..(b,d) |> f(a,_,c)`  becomes `f(a,b,c,d)`

**Comment by jorsn** [comment]
I thought, we have partial function application like `F(arg1)`only with `with`, e.g. `F.with(arg1)`? If you want to get rid of `.with`, you could introduce `F(arg1) = arg2 => F(arg1, arg2)` and thus, `F(arg1)(arg2) = F(arg1, arg2)`, which is just an implicit conversion between [curried and uncurried](https://en.wikipedia.org/wiki/Currying) functions.

I would prefer `f ° g = x => f(g(x))`, which really chains the functions. As a consequence, one could pass `f ° g` as an argument to another function, e.g. `h(f ° g)`, while `h(g |> f)` would mean `h(g(f))`, which is different. Having only reverse function application (`|>`) is much less powerful than function chaining! Evaluation order of `°` is from right to left, which is the usual thing for functions (think of, e.g. $(f \circ g) (x) = f(g(x))$ ). `(x, y) |> g` reminds me of the infamous reverse polish notation that, e.g., BibTeX style files use...

I think, having many parentheses is just a consequence of typst having curried functions and methods, instead of the Haskell style of uncurried functions and infix operators. That means, instead of writing `f a b c` and chaining `(g . h) x == g (h x)` as in Haskell, you write `f(a,b,c)` and the idiomatic way to chain functions would be something like `f.and(g)(x) = f(g(x))`. One might call the method `chain` instead of `and`. Of course, one usually has a limited number of infix operators, like arithmetic operators.

I guess, the curried syntax was chosen to look more familiar to the average academic without a background in functional programming, but it would be interesting to know @laurmaedje's view on this, since he is the main language designer, as I understand.
I think, this is a fundamental philosophical decision, and if you introduce many infix operators like |>, you might just as well introduce an uncurried syntax and rely on infix operators in general.

Therefore, I would introduce infix operators only for the most frequent needs, to stay rather consistent. Though, function chaining is probably  THE most frequent use in a functional language.

**Reply by astrale-sharp** [reply]
good points, didn't see it cause this discussion didn't trigger any notifications for me, sorry!

**Comment by Jollywatt** [comment]
Here’s an alternative idea for how to reduce clutter for nested functions: juxtaposed trailing argument syntax (JTA).

In typical code, initial arguments are commonly _options_ while the final argument is _content_, possibly nested in further function calls.

```rs
#align(center, text(size: 14pt, [Lorem Ipsum]))
```

Typst already formally prefers placing content _after_ normal arguments, supporting special syntax for the “arguments first, content last” pattern:

```rs
#align(center, text(size: 14pt)[Lorem Ipsum])
```

Keeping with this principle, why not simply allow:

```rs
#align(center)text(size: 14pt)[Lorem Ipsum]
```

I.e., two juxtaposed syntactic function calls are chained with right associativity.

It may appear peculiar at first sight (viewed in a vacuum, does `f(x)g(y)` mean `f(x, g(y))` or `g(f(x), y)`?) but in situ for conventional typst code it reads very naturally.

- JTA is not as flexible as a some proposed pipe operators, though it covers perhaps the most useful case.
- Outside-inside flow is preserved, unlike most versions of `|>`:
    ```rs
    Outermost...                                          ...Innermost
    #align(right,box(align(center, text(size: 14pt, [Lorem Ipsum]))))
    #align(right)box()align(center)text(size: 14pt)[Lorem Ipsum]
    ```
- To translate back and forth only involves (re)moving `)`s and `,`s.
- JTA is currently a syntax error in script mode, making it available in theory. The greediness of `#` must be adjusted, however — currently, we have:
    ```rs
    ┌───────────────┐ script mode
    #box(stroke: 1pt)text(fill: blue)[Hello]
                     └─────────────────────┘ content mode
    ````
    Cases where this is intended can be qualified by, e.g.,`Corner#hyph(2)case(s)`→ `Corner#(hyph(2))case(s)`
- This may save some jumping back and forth with parentheses, but I’m on the fence about it overall. At least `align(center, text(size: 14pt, [Lorem Ipsum]))` is simple and honest. Lisp users would certainly be fine with `))))`s!

**Reply by Jollywatt** [reply]
Afterthought: Perhaps we could use `:` instead of juxtiposition?

```rs
// current
align(right, box(align(center, text(size: 14pt, [Lorem Ipsum]))))
// proposed
align(right): box(): align(center): text(size: 14pt)[Lorem Ipsum]
```

This allows for multi-line chains:

```rs
// current
align(center, box(stroke: 1pt, text(size: 14pt, [Lorem Ipsum])))
// proposed
align(center):
    box(stroke: 1pt):
        text(size: 14pt)[Lorem Ipsum]
```

Spaces could be omitted for in-line code:
```rs
      ┌──────────────────────────────────────┐ script mode
Hello #box(stroke: 1pt):text(fill: red)[world]!
```

**Comment by SommerEngineering** [comment]
I don't see a problem with the syntax as it is 🤷‍♂️ :

```
function(arg1,other_func(arg2,arg3,last_func(arg4,arg5),arg6,arg7))
```

This is in line with the expectation of many developers, as this is how it would look in many programming languages. In my view, the alternatives proposed here are all worse than the current state because they are unusual and unexpected.

When the nesting of functions becomes too complex, I can do the same thing in Typst that I can do in corresponding programming languages -- store the results of the other calls:

```
#let last_result = last_func(arg4,arg5)
#let other_result = other_func(arg2,arg3,last_result,arg6,arg7)
function(arg1,other_result)
```

In my opinion, this ensures the best readability 😀

**Comment by Enivex** [comment]
> I don't see a problem with the syntax as it is 🤷‍♂️ :
> 
> ```
> function(arg1,other_func(arg2,arg3,last_func(arg4,arg5),arg6,arg7))
> ```
> 
> This is in line with the expectation of many developers, as this is how it would look in many programming languages. In my view, the alternatives proposed here are all worse than the current state because they are unusual and unexpected.
> 
> When the nesting of functions becomes too complex, I can do the same thing in Typst that I can do in corresponding programming languages -- store the results of the other calls:
> 
> ```
> #let last_result = last_func(arg4,arg5)
> #let other_result = other_func(arg2,arg3,last_result,arg6,arg7)
> function(arg1,other_result)
> ```
> 
> In my opinion, this ensures the best readability 😀 

Regardless of your opinion of piping, it wouldn't replace the existing syntax, merely complement it.

**Comment by astrale-sharp** [comment]
It's been decided that the new syntax introduction wasn't worth it (at least for now, might change if amazingly  good arguments are proposed)
